### DISCUSSION TITLE
Restriction on special characters in variable names?

### DISCUSSION THREAD

**Discussion by RedPxnda** [ezylang/EvalEx#383] [Q&A] [answered]
I've tried to use special characters like `]`, `}`, `@`, etc. in variable names, but I get a ParseException with a message saying `Undefined operator '<whatever the character is>'`.

I entirely get why this error is thrown, (issues with operators like >, =, -, +, etc.) but is there any way to work around non-operator characters? I really want to have my variables marked with some indicator that they're variables.

**Comment by uklimaschewski** [comment, marked as answer]
Hi, variable names must start with a letter or an underscore. So the underscore is the only special character that is allowed for a variable:
```java
    Expression expression = new Expression("_a + _b");
    expression.with("_a", 2).and("_b", 3).evaluate();
```
With some changes, a feature could be added to allow configuring additional characters, e.g. the _$_ character is a common special character denoting variables, e.g. this could look like this:
```java
    Expression expression = new Expression("$a + $b");
    expression.with("$a", 2).and("$b", 3).evaluate();
```

If you want to contribute, then you are welcome. :-)

**Comment by RedPxnda** [comment]
Thanks for the quick response!

I could make a PR real quick to make this change for dollar signs, so I'll get work onto that now.

A little off topic, but I was also wanting a couple changes for personal use(assuming that most of these changes are not ones you would want to have on the actual project) (eg. ability to use ' instead of " for string literals, possibly even some sort of token registry thing), so would I be allowed to fork this project?

**Reply by uklimaschewski** [reply]
This is an open source project, of course you can fork it and submit pull requests.
Here are some tips to get your PR merged quickly:
- Follow code style rules, you can automatically format your code using `mvn spotless:apply`.
- Add meaningful unit tests, using AssertJ for assertions. Merging also requires a 100% code coverage.
- Adopt the documentation, if needed.
- EvalEx uses semantic versioning. If you add a new feature, then make a pull request to the next unreleased minor version branch. Here this is the 3.1.x branch at the moment.

I would like to have two configuration properties for this feature:
- `List<Character> additionalVariableStartCharacters = Collections.singletonList('_')`
- `List<Character> additionalVariableCharacters = Collections.singletonList('_')`

Instead of hard coding the '_' underscore as allowed character for variable start and inside variables, I would like to have separate configurable lists for both. So here one could add the '$' for variable start (and even inside a variable name, if wanted).

Code pointers:
https://github.com/ezylang/EvalEx/blob/df8eeeb85e2402bb88784729c497d9b3cfcfa27d/src/main/java/com/ezylang/evalex/parser/Tokenizer.java#L543
https://github.com/ezylang/EvalEx/blob/df8eeeb85e2402bb88784729c497d9b3cfcfa27d/src/main/java/com/ezylang/evalex/parser/Tokenizer.java#L547

For the option to use single quotes instead of double quotes, please create a separate pull request. Also here, this should be configurable so it does not break backward compatibility.
